bin/android-functions.sh

```diff
@@ -4,6 +4,268 @@ declare root="$(cd "$(dirname "$(dirname "${BASH_SOURCE[0]}")")" && pwd)"
+declare ANDROID_STUDIO_PAGE_URI="https://developer.android.com/studio"
+declare JDK_VERSION="19.0.2"
+declare JDK_URI_TEMPLATE="https://download.java.net/java/GA/jdk$JDK_VERSION/fdb695a9d9064ad6b064dc6df578380c/7/GPL/openjdk-$JDK_VERSION""_""{os}-{arch}_bin.{format}"
+declare GRADLE_URI_TEMPLATE="https://services.gradle.org/distributions/gradle-8.0.2-bin.zip"
+
+# TODO(mribbons): ubuntu / apt libs: apt-get install libc6:i386 libncurses5:i386 libstdc++6:i386 lib32z1 libbz2-1.0:i386
+
+function get_android_default_search_paths() {
+  local host="$(host_os)"
+
+  # Prepopulate search paths with any known paths so that they can be verified quickly
+  ANDROID_HOME_SEARCH_PATHS=()
+  if [ -n "$ANDROID_HOME" ]; then
+    ANDROID_HOME_SEARCH_PATHS+=("$ANDROID_HOME")
+  fi
+
+  ANDROID_HOME_SEARCH_PATHS+=("${DEFAULT_ANDROID_HOME["$host"]}")
+
+  ANDROID_SDK_MANAGER_SEARCH_PATHS=()
+  if [ -n "$ANDROID_SDK_MANAGER" ]; then
+    ANDROID_SDK_MANAGER_SEARCH_PATHS+=("$(dirname "$ANDROID_SDK_MANAGER")")
+  fi
+
+  ANDROID_SDK_MANAGER_SEARCH_PATHS+=("${ANDROID_SDK_MANAGER_DEFAULT_SEARCH_PATHS[@]}")
+
+  JAVA_HOME_SEARCH_PATHS=()
+  if [ -n "$JAVA_HOME" ]; then
+    JAVA_HOME_SEARCH_PATHS+=("$JAVA_HOME")
+  fi
+
+  GRADLE_SEARCH_PATHS=()
+
+  if [ -n "$GRADLE_HOME" ]; then
+    GRADLE_SEARCH_PATHS+=("$GRADLE_HOME")
+  fi
+  
+  if [ -n "$HOME" ]; then
+    GRADLE_SEARCH_PATHS+=("$HOME/.gradle")
+  fi
+
+  # Only attempt default homes if $ANDROID_HOME not defined
+  if [[ "$host" = "Darwin"  ]]; then
+    JAVA_HOME_SEARCH_PATHS+=("$HOME/.local/bin")
+    JAVA_HOME_SEARCH_PATHS+=("$HOME/Applications")
+    JAVA_HOME_SEARCH_PATHS+=("$HOME/homebrew")
+    JAVA_HOME_SEARCH_PATHS+=("/Applications")
+  elif [[ "$host" = "Linux"  ]]; then
+    ANDROID_HOME_SEARCH_PATHS+=("$HOME")
+    JAVA_HOME_SEARCH_PATHS+=("$HOME/.local/bin")
+    # Java should be unpacked alongside Android Studio
+    JAVA_HOME_SEARCH_PATHS+=("$HOME")
+    JAVA_HOME_SEARCH_PATHS+=("/opt")
+  elif [[ "$host" = "Win32"  ]]; then
+    JAVA_HOME_SEARCH_PATHS+=("$LOCALAPPDATA\Programs")
+    JAVA_HOME_SEARCH_PATHS+=("$PROGRAMFILES")
+    GRADLE_SEARCH_PATHS+=("$LOCALAPPDATA\Programs")
+  fi
+
+  export ANDROID_HOME_SEARCH_PATHS
+  export JAVA_HOME_SEARCH_PATHS
+  export GRADLE_SEARCH_PATHS
+}
+
+function test_javac_version() {
+  local javac=$1
+  local target_version=$2
+
+  jc_v="$("$javac" --version 2>/dev/null)"; r=$?
+  if [[ "$r" != "0" ]]; then
+    return $r
+  fi
+
+  jc_v="${jc_v/javac\ /}"
+  write_log "v" "Comparing $javac version: "$(version "$jc_v") "$(version "$target_version")"
+  if [ "$(version "$jc_v")" -lt "$(version "$target_version")" ]; then
+    return 1
+  fi
+
+  return 0
+}
+
+function get_android_paths() {
+  get_android_default_search_paths
+
+  local _ah
+  local _sdk
+  local _jh
+  local _gh
+
+  for android_home_test in "${ANDROID_HOME_SEARCH_PATHS[@]}"; do
```

Comment:
`android_home_test` and `sdk_man_test` are leaked

Reply:
Is the solution to just declare locally above the loops?

Reply:
yeah, above with the rest should work